### PR TITLE
Chore: update artifact actions

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
The `master` tag was pointing to a very old build that sometimes failed (the "master" branch is now "main"). More appropriately pointing to a version now. Also, changed to avoid compressing artifacts that are already compressed.

In the process, also updated other actions including setup-go, which comes with its own caching for Go deps so we don't need the custom cache tasks

### CODE REVIEW

**Review by artursouza — APPROVED** [review]

**Review by daixiang0 — APPROVED** [review]
LGTM

**Comment by artursouza** [comment]
There are conflicts now.

**Comment by ItalyPaleAle** [comment]
> There are conflicts now.

Fixed

**Review by daixiang0 — DISMISSED** [review]
LGTM

File: .github/workflows/dapr.yml [hunk 1]
```diff
@@ -400,27 +303,30 @@ jobs:
           make release-flavor
         shell: bash
       - name: upload artifacts
-        uses: actions/upload-artifact@master
+        uses: actions/upload-artifact@v4
+        # Avoid publishing duplicate Windows artifacts, which will cause an error
+        if: matrix.windows_version != '1809'
```
Comment (ItalyPaleAle) [inline .github/workflows/dapr.yml:308]: This is key to fixing the builds right now: https://github.com/dapr/dapr/actions/runs/7534402274/job/20516204786

(The binaries are the same on 1809 and LTSC2022)

**Review by yaron2 — APPROVED** [review]

**Pull request merged**